tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -0,0 +1,236 @@
+    Shape weightsShapeStatic = sampleWeights.shape();
+    int weightsRankStatic = weightsShapeStatic.numDimensions();
+
+    Operand<TInt32> valuesShape = tf.shape(values);
+    Operand<TInt32> valuesRank = tf.rank(values);
+    Shape valuesShapeStatic = values.shape();
+    int valuesRankStatic = valuesShapeStatic.numDimensions();
+
+    if (weightsRankStatic != -1 && valuesRankStatic != -1) {
+      if (weightsRankStatic == 0) {
+        return tf.withSubScope("static_scalar_check_success")
+            .withControlDependencies(Collections.EMPTY_LIST)
+            .noOp();
+      }
+      if (weightsRankStatic != valuesRankStatic) {
+        throw new IllegalArgumentException(
+            String.format(
+                "%s values.rank=%d. weights.rank=%d.  values.shape=%s. weights.shape=%s.",
+                ASSERT_BROADCAST_ERROR_PREFIX,
+                valuesRankStatic,
+                weightsRankStatic,
+                valuesShapeStatic.toString(),
+                weightsShapeStatic.toString()));
+      }
+
+      for (int i = 0; i < valuesRankStatic; i++) {
+        if (valuesShapeStatic.size(i) != weightsShapeStatic.size(i)) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "%s Mismatch at dim %d. values.shape=%s weights.shape=%s.",
+                  ASSERT_BROADCAST_ERROR_PREFIX,
+                  i,
+                  valuesShapeStatic.toString(),
+                  weightsShapeStatic.toString()));
+        }
+      }
+      return tf.withSubScope("static_dims_check_success")
+          .withControlDependencies(Collections.EMPTY_LIST)
+          .noOp();
+    }
+    // Dynamic checks.
+    Operand<TBool> is_scalar = tf.math.equal(weightsRank, tf.constant(0));
+    List<Operand<?>> data =
+        Arrays.asList(
+            tf.constant(ASSERT_BROADCAST_ERROR_PREFIX),
+            tf.constant("weights.shape="),
+            weightsShape,
+            tf.constant("values.shape="),
+            valuesShape,
+            tf.constant("is_scalar="),
+            is_scalar);
+
+    Operand<TBool> isValidShape =
+        tf.select(
+            is_scalar,
+            is_scalar,
+            hasValidNonscalarShape(tf, weightsRank, weightsShape, valuesRank, valuesShape));
+
+    return tf.withSubScope("broadcastWeights-dynamic").assertThat(isValidShape, data);
+  }
+
+  /**
+   * Gets an operand that tests if the shapes have the same rank and valid dimensions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param weightsRank the operand for the rank of the sample weights
+   * @param weightsShape the operand for the shape of the sample weights
+   * @param valuesRank the operand for the rank of the sample weights
+   * @param valuesShape the operand for the shape of the sample weights
+   * @param <T> the data type for the operands
+   * @return a boolean operand to determine if the Shape is scalar or not.
+   */
+  private static <T extends TNumber> Operand<TBool> hasValidNonscalarShape(
```

Comment:
is was `has_valid_nonscalar_shape` in Python. Since this is private, the name doesn't matter to me. So if you think `canBroadcastNonscalarShapes` is a better name, then let's change it.

Reply:
`canBroadcastNonscalarShapes` does seem better to me. I like the clue that this is part of the `canBroadcast` collection of methods, whereas the name `hasValidNonscalarShape` doesn't really mean anything from a standing start.

Reply:
OK changed to `canBroadcastNonscalarShapes`

Reply:
👍 done